mlx/backend/metal/cholesky.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright © 2024 Apple Inc.
+
+#include "mlx/backend/metal/mps/cholesky.h"
+#include "mlx/backend/metal/copy.h"
+#include "mlx/backend/metal/device.h"
+#include "mlx/backend/metal/kernels.h"
+#include "mlx/backend/metal/kernels/defines.h"
+#include "mlx/backend/metal/kernels/steel/gemm/params.h"
+#include "mlx/backend/metal/matmul.h"
+#include "mlx/backend/metal/mps/gemm.h"
+#include "mlx/backend/metal/utils.h"
+#include "mlx/primitives.h"
+#include "mlx/utils.h"
+
+namespace mlx::core {
+
+void Cholesky::eval_gpu(const std::vector<array>& inputs, array& out) {
+  auto& s = stream();
+  auto& d = metal::device(s.device);
+
+  const auto& a = inputs.at(0);
+  if (a.dtype() != float32) {
+    throw std::runtime_error("Cholesky::eval_gpu only supports float32.");
+  }
+
+  out.set_data(allocator::malloc_or_wait(out.nbytes()));
+
+  MPS::DataType mps_dtype = MPS::DataTypeFloat32;
+
+  const int num_rows = a.shape(-2);
+  const int row_nbytes = a.shape(-1) * sizeof(float);
+  const int num_matrices = a.size() / (num_rows * num_rows);
+
+  const auto mat_desc = MPS::MatrixDescriptor::matrixDescriptor(
+      /* rows = */ num_rows,
+      /* columns = */ num_rows,
+      /* rowBytes = */ row_nbytes,
+      /* dataType = */ mps_dtype);
+
+  d.end_encoding(s.index);
+
+  // The kernel does not support batching, so we have to loop explicitly.
+  for (int i = 0; i < num_matrices; i++) {
+    size_t offset = i * num_rows * row_nbytes;
+    auto a_buf = static_cast<const MTL::Buffer*>(a.buffer().ptr());
+    auto a_mat = MPS::Matrix::alloc()->init(a_buf, offset, mat_desc);
+
+    auto out_buf = static_cast<MTL::Buffer*>(out.buffer().ptr());
+    auto out_mat = MPS::Matrix::alloc()->init(out_buf, offset, mat_desc);
+
+    auto kernel = MPS::MatrixDecompositionCholesky::alloc()->init(
+        d.mtl_device(), !upper_, num_rows);
+
+    auto command_buffer = d.get_command_buffer(s.index);
+    kernel->encodeToCommandBuffer(command_buffer, a_mat, out_mat, nullptr);
+    command_buffer->addCompletedHandler(
+        [a_mat, out_mat, kernel](MTL::CommandBuffer*) mutable {
+          a_mat->release();
+          out_mat->release();
+          kernel->release();
+        });
+  }
+
+  // Zero out the unwanted part of the output.
```

Comment:
That's interesting that it's needed. Is it because the memory is uninitialized or does the MPS kernel actually write some stuff to the 0 triangle? If it's the former, probably simpler to just zero the memory initially.

Reply:
It's used as scratch space.